applications_superstaq/check/all_.py

```diff
@@ -17,12 +16,7 @@
 )
 
 
-@check_utils.extract_file_args
-def run(
-    *args: str,
-    files: Optional[Iterable[str]] = None,
-    parser: argparse.ArgumentParser = check_utils.get_file_parser()
-) -> int:
+def run(*args: str, parser: argparse.ArgumentParser = check_utils.get_file_parser()) -> int:
```

Comment:
should probably use something like
```suggestion
def run(*args: str, parser: Optional[argparse.ArgumentParser] = None) -> int:
    if parser is None:
        parser = check_utils.get_file_parser()
```
because ArgumentParser() is mutable

(ditto elsewhere)

Reply:
I don't think I understand the issue.  Can you explain?

Reply:
>  Can you explain?

there are two issues i can think of:

first is pretty insignificant: all the `get_file_parser()` calls will happen as soon as the submodules are first imported instead of when a particular `run()` is called, which means we're doing the unnecessary work of instantiating all the parsers in check/*.py even if only one is actually being used (however (i think) these should be pretty lightweight so probably not a real problem in this case) 

second: because the resulting parser is mutable it's easy to end up with strange side effects. This is why pylist normally throws a "dangerous-default-value" error for mutable default arguments (apparently this only occurs for built-in types, which I suppose makes sense because pylint isn't a type parser). In this case, if you call `all_.run()` more than once it will raise and `ArgumentError: conflicting option string...` because the second call will try to add the same arguments to the default parser a second time. It's true that this shouldn't come up in normal use (because when run as a script `run()` won't be called more than once), but it's not portable and i can easily imagine it causing issues e.g. when we start adding checks for the checks

Reply:
Follow-up: should I rename `parser` to `_parser` to emphasize the fact that nobody calling `run` should actually pass in a parser?

Reply:
> second issue: appears to not be the case.  Example: if I add a `return 0` after parsing args in `pytest_.py`, I can try `run(); run()`, which executes without errors.  I can also, say, `print(parser.description)` before and after defining it, and the prints behave as desired in both runs.

it doesn't seem to be an issue in pytest_.py, but it is in all_.py. For example if i add a line:
```
if __name__ == "__main__":
    print(run(*sys.argv[1:]))
    exit(run(*sys.argv[1:]))
```
the second `run()` raises `argparse.ArgumentError: argument -f: conflicting option string: -f`

Reply:
I see, I think it's an issue in `all_.py` because everywhere else I actually replace the parser with a new one inside the `enable_incremental` wrapper.  In any case, moot point now I think.

Reply:
> why decorator

i do really appreciate the decorator aesthetics, but i guess i think it ultimately ends up making both the *_ scripts and check_utils somewhat opaque/harder to follow: it kind of takes away the from "magically enable" functionality if you also have to add an extra unused argument in the function signature, and it's not really obvious what it's doing if you just read e.g. pytest_.py (i definitely wouldn't expect it to be using `inspect` to extract and update the default value of a function argument, which afaik isn't a typical practice in python)

so i guess imo making it a function would make it much simpler more obvious to future developers/maintainers what it's actually doing. But definitely ok with being overruled here; will approve (pending a few nits) if you think it should stay as is

Reply:
[done](https://github.com/SupertechLabs/applications-superstaq/pull/92#issuecomment-1182534555)